docs/variance-defaulting.md

```diff
@@ -0,0 +1,375 @@
+  my_box.class.new('')
+  ```
+
+  If `my_box` is `Box[Integer]`, there's no reason why `my_box.class.new` has to
+  create another `Box[Integer]`--making a `Box[String]` is fine!
+
+  But in Sorbet's CFG, that snippet looks like this:
+
+  ```ruby
+  <tmp>$1 = my_box.class
+  <tmp>$2 = <tmp>$1.new('')
+  ```
+
+  and Sorbet has no way of knowing whether `<tmp>$1` was a constant literal or
+  an expression. (Wildly, Sorbet allows `my_box.class[String].new('')` to make a
+  well-typed `Box[String]` instance).
+
+**TL;DR**: It's unavoidable for Sorbet to have to allow bare generics in places.
+
+However, completely bare generics are bad, because then we have no types to
+substitute for generic types when used in method signatures, etc.
+
+So Sorbet has default rules to silently convert bare generic classes to applied
+generic classes in places where it's unavoidable.
+
+## The defaulting rules
+
+So we've established that bare generic classes are bad, and that bare generic
+classes are also unavoidable. Defaulting rules are how we make the best of a bad
+situation.
+
+The defaulting rules are implemented by `ClassOrModule::externalType()`, which
+is computed once per Symbol during resolver and cached, so the actual
+computation happens in [`unsafeComputeExternalType`]. (It's not actually unsafe
+anymore because of a change jvilk made in [this commit][175ce0f1]. It used to do
+lurky things with atomics and racy mutable access).
+
+The rules here might be out of date w.r.t. the implementation, and if they are
+please edit this doc. But at a high level:
+
+- If a type_member is a "legacy" generic, which is basically `Array` and a few
+  other classes (but importantly, **not** every stdlib generic class), it
+  unsafely defaults to `T.untyped` no matter what. (This was just a hack to get
+  things to type check when rolling out Sorbet.)
+
+- If a type member is `fixed`, default it to whatever the type member is fixed
+  to, regardless of variance.
+
+- Otherwise, we're going to have to use the variance:
+
+  - Invariant? -> `T.untyped`
+  - Covariant? -> the `upper` bound (the default upper bound is `T.anything`)
+  - Contravariant? -> the `lower` bound (the default lower bound is
+    `T.noreturn`)
+
+The most common question is then: why are invariant type members defaulted to
+`T.untyped`? It's because we can't get away with using `upper` nor `lower`.
+
+Consider this setup:
+
+```ruby
+class Parent; end
+class Child < Parent; end
+```
+
+In this setup, `T.class_of(Child) <: T.class_of(Parent)`, which is what users
+expect. Now consider if we made `Parent` and `Child` generic:
+
+```ruby
+class Parent
+  extend T::Generic
+  Elem = type_template { {upper: Numeric} }
+end
+class Child < Parent
+  extend T::Generic
+  Elem = type_template { {upper: Integer} }
+end
+```
+
+It no longer makes sense to ask whether `T.class_of(Child) <:
+T.class_of(Parent)`, because those are bare generics. But both `upper` and
+`lower` are hard to stomach as defaults.
```

Comment:
It feels to me like there is a step missing here between the two sentences, which is that if `Child` or `Parent` occur as bare generics (even though those are bad), we need to pick *something* to provide for `Elem`.  WDYT?